src/adventure/adventurewidget.cpp

```diff
@@ -0,0 +1,165 @@
+    auto font = new QFont();
+    font->fromString(settings.font); // needed fromString() to extract PointSize
+    blockCharFormat.setFont(*font);
+    m_journalTextCursor->setBlockCharFormat(blockCharFormat);
+
+    addJournalEntry(DEFAULT_MSG);
+
+    auto *layout = new QVBoxLayout(this);
+    layout->setAlignment(Qt::AlignTop);
+    layout->setContentsMargins(0, 0, 0, 0);
+    layout->setSpacing(0);
+    layout->addWidget(m_journalTextEdit);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_accomplishedTask,
+            this,
+            &AdventureWidget::slot_onAccomplishedTask);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_achievedSomething,
+            this,
+            &AdventureWidget::slot_onAchievedSomething);
+
+    connect(&m_adventureTracker, &AdventureTracker::sig_died, this, &AdventureWidget::slot_onDied);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_gainedLevel,
+            this,
+            &AdventureWidget::slot_onGainedLevel);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_killedMob,
+            this,
+            &AdventureWidget::slot_onKilledMob);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_receivedHint,
+            this,
+            &AdventureWidget::slot_onReceivedHint);
+}
+
+void AdventureWidget::slot_onAccomplishedTask(double xpGained)
+{
+    // Only record accomplishedTask if it actually has associated xp to avoid
+    // spam, since sometimes it co-triggers with achievemnt and can be redundant.
+    if (xpGained > 0.0) {
+        auto msg = QString(ACCOMPLISH_MSG) + QString(XP_SUFFIX).arg(formatXPGained(xpGained));
+        addJournalEntry(msg);
+    }
+}
+
+void AdventureWidget::slot_onAchievedSomething(const QString &achievement, double xpGained)
+{
+    auto msg = QString(ACHIEVE_MSG).arg(achievement);
+
+    if (xpGained > 0.0)
+        msg = msg + QString(XP_SUFFIX).arg(formatXPGained(xpGained));
+
+    addJournalEntry(msg);
+}
+
+void AdventureWidget::slot_onDied(double xpLost)
+{
+    auto msg = QString(DIED_MSG);
+
+    if (xpLost < 0.0)
+        msg = msg + QString(XP_SUFFIX).arg(formatXPGained(xpLost));
+
+    addJournalEntry(msg);
+}
+
+void AdventureWidget::slot_onGainedLevel()
+{
+    addJournalEntry(QString(GAINED_LEVEL_MSG));
+}
+
+void AdventureWidget::slot_onKilledMob(const QString &mobName, double xpGained)
+{
+    auto msg = QString(KILL_TROPHY_MSG).arg(mobName);
+
+    if (xpGained > 0.0) {
+        msg = msg + QString(XP_SUFFIX).arg(formatXPGained(xpGained));
+    } else {
+        // When player gets XP from multiple kills on the same heartbeat, like
+        // frequently happens with quake xp, then the first mob gets all the XP
+        // attributed and the others are zero. No way to solve this given
+        // current "protocol" of how MUME exposes information.
+        msg = msg + QString(XP_SUFFIX).arg("?");
```

Comment:
MUME's engine event loop works in 250ms pulses so you could assume it's split equally across the current pulse. That's probably better than attributing it to one source (but more code complexity).

Reply:
So the tricky part is: how do I detect the beginning and end of a heartbeat reliably? Blank lines is probably the only protocol, or some combination of blank lines and prompts.

I loaded up several years of MUME logs into a DB, and there is a long tail of obscure scenarios where things can get a little jumbled and lots of messages can appear together, or even mismatched, and tons of corner cases (like your lantern oil runs out in the middle of a fight). The logic I have above (share of experience + R.I.P. within the space of a few lines) is robust against these (in my testing, at least) and covers something like 99.x% of cases. Going after robust, "full heartbeat" parsing is a significant task.

But ultimately, just in terms of personal preference, I'm not sure "smooth average across heartbeat" is that much better. If you're quaking a cave full of snakes or bats, it will probably be more accurate, but if there is any mob level disparity it will still be off. Also, one nice aspect of attributing to one source is that it shows you how much you get for the whole quake at a glance (like "whoa, I just got 11k for quaking that room of snakes").

My strong preference would be to keep the parsing logic simple for initial release and then if anyone actually uses this thing, and likes it, I can consider a more robust heartbeat parsing framework. I worry the parsing complexity is just not worth the risk now

Reply:
Yeah, logs won't have it. The primary chaos factors to be aware of are network lag and the game engine lag. We could probably do a fairly decent estimate but this goes back to the earlier model you had with a sliding window. I agree with your assessment that it's not necessary for the MVP :-)